Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+                .filter { it.isPublicKeyRegistered && it.isPending }
+                .map { cert ->
+                    try {
+                        RefreshResult(obtainCertificate(cert))
+                    } catch (e: Exception) {
+                        Timber.tag(TAG).e(e, "Failed to retrieve test certificate for %s", cert)
+                        RefreshResult(cert, e)
+                    }
+                }
+
+            mutate {
+                refreshedCerts
+                    .filter { it.error == null }
+                    .map { it.certificate }
+                    .forEach { this[it.identifier] = it }
+            }
+        }
+
+        internalData.updateBlocking {
+            val certs = values.filter { workedOnIds.contains(it.identifier) }
+
+            mutate {
+                certs.forEach {
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = false)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = false)
+                    }
+                }
+            }
+        }
+
+        return refreshResults
+    }
+
+    private suspend fun registerPublicKey(
+        cert: TestCertificateContainer
+    ): TestCertificateContainer {
+        return try {
+            Timber.tag(TAG).d("registerPublicKey(cert=%s)", cert)
+
+            if (cert.isPublicKeyRegistered) {
+                Timber.tag(TAG).d("Public key is already registered for %s", cert)
+                return cert
+            }
+
+            val rsaKeyPair = rsaKeyPairGenerator.generate()
+
+            withContext(dispatcherProvider.IO) {
+                certificateServer.registerPublicKeyForTest(
+                    testRegistrationToken = cert.registrationToken,
+                    publicKey = rsaKeyPair.publicKey,
+                )
+            }
+            Timber.tag(TAG).i("Public key successfully registered for %s", cert)
+
+            val nowUTC = timeStamper.nowUTC
+
+            when (cert.type) {
+                CoronaTest.Type.PCR -> (cert as PCRCertificateContainer).copy(
+                    publicKeyRegisteredAt = nowUTC,
+                    rsaPublicKey = rsaKeyPair.publicKey,
+                    rsaPrivateKey = rsaKeyPair.privateKey,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> (cert as RACertificateContainer).copy(
+                    publicKeyRegisteredAt = nowUTC,
+                    rsaPublicKey = rsaKeyPair.publicKey,
+                    rsaPrivateKey = rsaKeyPair.privateKey,
+                )
+            }
+        } catch (e: Exception) {
+            Timber.tag(TAG).e("Failed to register public key for %s", cert)
+            throw e
+        }
+    }
+
+    private suspend fun obtainCertificate(
+        cert: TestCertificateContainer
+    ): TestCertificateContainer {
```

Comment:
Try to obtain the actual certificate. PublicKey registration and certificate retrieval are two steps, because if we manage to register our public key, but fail to get the certificate, we are still one step further.

Reply:
It would be nice if you can this comment as a Kotlin comment in code ? PR is gone, comment in there for later